benchmarks/graphsim.py

```diff
@@ -0,0 +1,168 @@
+"""
+
+# %%
+# Firstly, let us import relevant modules:
+
+from copy import copy
+from time import perf_counter
+
+import matplotlib.pyplot as plt
+import numpy as np
+
+from graphix import Circuit
+
+# %%
+# Next, define a function to generate random Clifford circuits.
+
+
+def genpair(n_qubits, count, rng):
+    pairs = []
+    for i in range(count):
+        choice = [j for j in range(n_qubits)]
+        x = rng.choice(choice)
+        choice.pop(x)
+        y = rng.choice(choice)
+        pairs.append((x, y))
+    return pairs
+
+
+def random_clifford_circuit(nqubits, depth, seed=42):
+    rng = np.random.default_rng(seed)
+    circuit = Circuit(nqubits)
+    gate_choice = list(range(5))
+    for _ in range(depth):
+        for j, k in genpair(nqubits, 2, rng):
+            circuit.cnot(j, k)
+        for j in range(nqubits):
+            k = rng.choice(gate_choice)
+            if k == 0:  # H
+                circuit.h(j)
+            elif k == 1:  # S
+                circuit.s(j)
+            elif k == 2:  # X
+                circuit.x(j)
+            elif k == 3:  # Z
+                circuit.z(j)
+            elif k == 4:  # Y
+                circuit.y(j)
+            else:
+                pass
+    return circuit
+
+
+# %%
+# We generate a set of random Clifford circuits with different widths.
+
+DEPTH = 3
+test_cases = [i for i in range(2, 300, 10)]
+graphix_patterns = {}
+
+for i in test_cases:
+    circuit = random_clifford_circuit(i, DEPTH)
+    pattern = circuit.transpile()
+    pattern.standardize()
+    nodes, edges = pattern.get_graph()
+    graphix_patterns[i] = (circuit, pattern, len(nodes))
+
+
+# %%
+# We then run simulations.
+# First, we run the pattern optimization using networkx.
+networkx_time = []
+networkx_node = []
+
+for width, (circuit, pattern, num_nodes) in graphix_patterns.items():
+    pattern_copy = copy(pattern)
+    start = perf_counter()
+    pattern_copy.perform_pauli_measurements()
+    end = perf_counter()
+    networkx_node.append(num_nodes)
+    print(f"width: {width}, number of nodes: {num_nodes}, depth: {DEPTH}, time: {end - start}")
+    networkx_time.append(end - start)
+
+
+# %%
+# Next, we run the pattern optimization using rustworkx.
+rustworkx_time = []
+rustworkx_node = []
+
+for width, (circuit, pattern, num_nodes) in graphix_patterns.items():
+    pattern_copy = copy(pattern)
+    start = perf_counter()
+    pattern_copy.perform_pauli_measurements(use_rustworkx=True)
+    end = perf_counter()
+    rustworkx_node.append(num_nodes)
+    print(f"width: {width}, number of nodes: {num_nodes}, depth: {DEPTH}, time: {end - start}")
+    rustworkx_time.append(end - start)
+
+# %%
+# For reference, we also perform Anders and Briegel's stabilizer circuit simulator based on
+# graph state representation.
+# See https://arxiv.org/abs/quant-ph/0504117v2 or https://github.com/libtangle/graph-state for more details.
+
+from graph_state import GraphState as ABGraphState
```

Comment:
Can we include it in the requirements.txt?
Ref. https://stackoverflow.com/questions/16584552/how-to-state-in-requirements-txt-a-direct-github-source